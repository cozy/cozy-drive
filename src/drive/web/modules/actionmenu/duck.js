import { combineReducers } from 'redux'

const SHOW_ACTIONMENU = 'SHOW_ACTIONMENU'
const HIDE_ACTIONMENU = 'HIDE_ACTIONMENU'

// reducers
const isVisible = (state = false, action) => {
  if (action.meta && action.meta.hideActionMenu) {
    return false
  }
  switch (action.type) {
    case SHOW_ACTIONMENU:
      return true
    case HIDE_ACTIONMENU:
      return false
    default:
      return state
  }
}

const actionable = (state = null, action) => {
  if (action.meta && action.meta.hideActionMenu) {
    return null
  }
  switch (action.type) {
    case SHOW_ACTIONMENU:
      return action.id || null
    case HIDE_ACTIONMENU:
      return null
    default:
      return state
  }
}

export default combineReducers({
  isVisible,
  actionable
})

// selectors
export const getActionableId = state => state.actionmenu.actionable
export const isMenuVisible = state => state.actionmenu.isVisible

// actions
export const showActionMenu = id => ({
  type: SHOW_ACTIONMENU,
  id
})

export const hideActionMenu = () => ({
  type: HIDE_ACTIONMENU
})
