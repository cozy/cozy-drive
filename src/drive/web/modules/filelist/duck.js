const SHOW_NEW_FOLDER_INPUT = 'SHOW_NEW_FOLDER_INPUT'
const HIDE_NEW_FOLDER_INPUT = 'HIDE_NEW_FOLDER_INPUT'

export const showNewFolderInput = () => ({
  type: SHOW_NEW_FOLDER_INPUT
})

export const hideNewFolderInput = () => ({
  type: HIDE_NEW_FOLDER_INPUT
})

const initialState = {
  isTypingNewFolderName: false
}

const filelist = (state = initialState, action) => {
  switch (action.type) {
    case SHOW_NEW_FOLDER_INPUT:
      return { ...state, isTypingNewFolderName: true }
    case HIDE_NEW_FOLDER_INPUT:
      return { ...state, isTypingNewFolderName: false }
    default:
      return state
  }
}

export default filelist

export const isTypingNewFolderName = state =>
  state.filelist.isTypingNewFolderName
