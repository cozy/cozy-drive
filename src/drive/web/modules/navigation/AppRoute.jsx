/* global __TARGET__ */
import React from 'react'
import { Route, Redirect, IndexRoute } from 'react-router'

import Settings from 'drive/mobile/modules/settings/Settings'
import OnBoarding from 'drive/mobile/modules/onboarding/OnBoarding'

import Layout from 'drive/web/modules/layout/Layout'
import FileExplorer from './FileExplorer'
import FilesViewer from 'drive/web/modules/viewer/FilesViewer'
import FileOpenerExternal from 'drive/web/modules/viewer/FileOpenerExternal'

import {
  FolderContainer as Folder,
  RecentContainer as Recent,
  SharingsContainer as Sharings
} from 'drive/web/modules/drive'
import { Container as Trash } from 'drive/web/modules/trash'

const AppRoute = (
  <Route>
    <Route component={Layout}>
      <Redirect from="/files/:folderId" to="/folder/:folderId" />
      <Route component={FileExplorer}>
        <Redirect from="/" to="folder" />
        <Route path="folder" component={Folder}>
          <Route path=":folderId">
            <Route path="file/:fileId" component={FilesViewer} />
          </Route>
          <Route path="file/:fileId" component={FilesViewer} />
        </Route>
        <Route path="recent" component={Recent}>
          <Route path="file/:fileId" component={FilesViewer} />
        </Route>
        <Route path="sharings">
          <IndexRoute component={Sharings} />
          <Route path=":folderId" component={Folder}>
            <Route path="file/:fileId" component={FilesViewer} />
          </Route>
          <Route path="file/:fileId" component={FilesViewer} />
        </Route>
        <Route path="trash" component={Trash}>
          <Route path=":folderId">
            <Route path="file/:fileId" component={FilesViewer} />
          </Route>
          <Route path="file/:fileId" component={FilesViewer} />
        </Route>
      </Route>
      {__TARGET__ === 'mobile' && (
        <Route path="settings" component={Settings} />
      )}
      <Route path="file/:fileId" component={FileOpenerExternal} />
    </Route>
    {__TARGET__ === 'mobile' && (
      <Route path="onboarding" component={OnBoarding} />
    )}
  </Route>
)

export default AppRoute
