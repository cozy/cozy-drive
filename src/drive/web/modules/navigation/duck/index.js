export {
  default,
  isNavigating,
  getVisibleFiles,
  getFileById,
  getOpenedFolderId,
  getFolderIdFromRoute,
  getFolderUrl,
  getSort
} from './reducer'

export {
  openFiles,
  openRecent,
  openTrash,
  openFolder,
  sortFolder,
  fetchMoreFiles,
  fetchRecentFiles,
  addFile,
  updateFile,
  deleteFile,
  getFileDownloadUrl,
  uploadFiles,
  createFolder,
  trashFiles,
  downloadFiles,
  exportFilesNative,
  openFileWith
} from './actions'
