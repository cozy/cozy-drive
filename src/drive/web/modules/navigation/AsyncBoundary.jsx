import { connect } from 'react-redux'
import { isNavigating } from 'drive/web/modules/navigation/duck'

const AsyncBoundary = ({ fetchStatus, isNavigating = false, children }) => {
  const fetchFailed = fetchStatus === 'failed'
  const fetchPending = fetchStatus === 'pending'
  const isLoading = fetchPending || isNavigating

  return children({ isLoading, isNavigating, isInError: fetchFailed })
}

const mapStateToProps = state => ({
  isNavigating: isNavigating(state),
  fetchStatus: state.view.fetchStatus
})

export default connect(mapStateToProps)(AsyncBoundary)
