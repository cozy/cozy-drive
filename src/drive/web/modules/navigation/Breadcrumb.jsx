/* global __TARGET__ cozy */
import React, { Component } from 'react'
import { withRouter } from 'react-router'
import { connect } from 'react-redux'
import classNames from 'classnames'
import Spinner from 'cozy-ui/react/Spinner'
import Icon from 'cozy-ui/react/Icon'
import { translate } from 'cozy-ui/react/I18n'
import { withBreakpoints } from 'cozy-ui/react'
import SharedDocuments from 'sharing/components/SharedDocuments'

import { ROOT_DIR_ID, TRASH_DIR_ID } from 'drive/constants/config'
import { openFolder, getFolderUrl } from 'drive/web/modules/navigation/duck'
import getFolderPath from './getFolderPath'

import styles from './breadcrumb.styl'

export const renamePathNames = (path, pathname, t) => {
  if (pathname === '/recent') {
    path.unshift({
      name: t('breadcrumb.title_recent')
    })
  } else if (pathname.match(/^\/sharings/)) {
    path.unshift({
      name: t('breadcrumb.title_sharings'),
      url: '/sharings'
    })
  }

  path.forEach(folder => {
    if (folder.id === ROOT_DIR_ID) {
      folder.name = t('breadcrumb.title_drive')
    } else if (folder.id === TRASH_DIR_ID) {
      folder.name = t('breadcrumb.title_trash')
    }
    if (!folder.name) folder.name = 'â€¦'
  })

  return path
}

export class Breadcrumb extends Component {
  state = {
    deployed: false
  }

  toggleDeploy = () => {
    this.state.deployed ? this.closeMenu() : this.openMenu()
  }

  openMenu() {
    this.setState({ deployed: true })
    document.addEventListener('click', this.handleClickOutside, true)
  }

  closeMenu() {
    this.setState({ deployed: false })
    document.removeEventListener('click', this.handleClickOutside, true)
  }

  handleClickOutside = e => {
    if (this.menu && !this.menu.contains(e.target)) {
      e.stopPropagation()
      this.closeMenu()
    }
  }

  render() {
    const {
      path,
      onBreadcrumbClick,
      opening,
      inlined,
      className = ''
    } = this.props
    const { deployed } = this.state

    if (!path) return false

    return (
      <div
        className={classNames(
          styles['fil-path-backdrop'],
          { [styles['deployed']]: deployed },
          { [styles['inlined']]: inlined },
          { [styles['mobile']]: __TARGET__ === 'mobile' },
          className
        )}
      >
        <h2
          data-test-id="path-title"
          className={styles['fil-path-title']}
          onClick={this.toggleDeploy}
          ref={ref => {
            this.menu = ref
          }}
        >
          {path.map((folder, index) => {
            if (index < path.length - 1) {
              return (
                <span
                  className={styles['fil-path-link']}
                  onClick={e => {
                    e.stopPropagation()
                    onBreadcrumbClick(folder)
                  }}
                  key={index}
                >
                  <span className={styles['fil-path-link-name']}>
                    {folder.name}
                  </span>
                  <Icon
                    icon="forward"
                    className={styles['fil-path-separator']}
                  />
                </span>
              )
            } else {
              return (
                <span
                  className={styles['fil-path-current']}
                  onClick={e => {
                    e.stopPropagation()
                    if (path.length >= 2) this.toggleDeploy()
                  }}
                  key={index}
                >
                  <span className={styles['fil-path-current-name']}>
                    {folder.name}
                  </span>
                  {path.length >= 2 && (
                    <span className={styles['fil-path-down']} />
                  )}

                  {opening && <Spinner />}
                </span>
              )
            }
          })}
        </h2>
      </div>
    )
  }
}

class RouterBreadCrumb extends Component {
  state = {
    opening: false
  }

  toggleOpening = () => {
    this.setState(state => ({ opening: !state.opening }))
  }

  navigateToFolder = folderId => {
    const { router, location, goToFolder, getFolderUrl } = this.props

    this.toggleOpening()

    goToFolder(folderId).then(() => {
      this.toggleOpening()
      router.push(getFolderUrl(folderId, location))
    })
  }

  navigateToPath = path => {
    const { router } = this.props
    router.push(path)
  }

  navigateTo = folder => {
    if (folder.id) this.navigateToFolder(folder.id)
    else this.navigateToPath(folder.url)
  }

  render() {
    const { path } = this.props
    const { opening } = this.state

    return (
      <Breadcrumb
        path={path}
        onBreadcrumbClick={this.navigateTo}
        opening={opening}
      />
    )
  }
}

export const PreviousButton = ({ onClick }) => (
  <button className={styles['fil-path-previous']} onClick={onClick} />
)

class RouterPreviousButton extends Component {
  navigateToFolder = folderId => {
    const { router, location, goToFolder, getFolderUrl } = this.props
    goToFolder(folderId).then(() => {
      router.push(getFolderUrl(folderId, location))
    })
  }

  navigateToPath = path => {
    const { router } = this.props
    router.push(path)
  }

  navigateBack = () => {
    const previousSegment = this.props.path[this.props.path.length - 2]
    if (previousSegment.id) this.navigateToFolder(previousSegment.id)
    else this.navigateToPath(previousSegment.url)
  }

  render() {
    return <PreviousButton onClick={this.navigateBack} />
  }
}

const MobileAwareBreadcrumb = props => {
  const { BarCenter, BarLeft } = cozy.bar

  return props.breakpoints.isMobile ? (
    <div>
      {props.path.length >= 2 && (
        <BarLeft>
          <RouterPreviousButton {...props} />
        </BarLeft>
      )}
      <BarCenter>
        <RouterBreadCrumb {...props} />
      </BarCenter>
    </div>
  ) : (
    <RouterBreadCrumb {...props} />
  )
}

const mapStateToProps = (state, ownProps) => ({
  path: renamePathNames(
    getFolderPath(
      state.view.displayedFolder,
      ownProps.location.pathname,
      ownProps.isPublic,
      ownProps.sharedDocuments
    ),
    ownProps.location.pathname,
    ownProps.t
  ),
  getFolderUrl
})

const mapDispatchToProps = dispatch => ({
  goToFolder: folderId => dispatch(openFolder(folderId))
})

const withSharedDocuments = Wrapped =>
  class withSharedDocumentsClass extends React.Component {
    render() {
      return (
        <SharedDocuments>
          {({ sharedDocuments }) => (
            <Wrapped sharedDocuments={sharedDocuments} {...this.props} />
          )}
        </SharedDocuments>
      )
    }
  }

export default withBreakpoints()(
  withRouter(
    translate()(
      withSharedDocuments(
        connect(
          mapStateToProps,
          mapDispatchToProps
        )(MobileAwareBreadcrumb)
      )
    )
  )
)
