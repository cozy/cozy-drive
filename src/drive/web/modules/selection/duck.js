import { combineReducers } from 'redux'
import { getFileById } from 'drive/web/modules/navigation/duck'
import get from 'lodash/get'

const getFileByIdFromCozyClient = (state, id) =>
  get(state, ['cozy', 'documents', 'io.cozy.files', id])

// constants
const SELECT_ITEM = 'SELECT_ITEM'
const UNSELECT_ITEM = 'UNSELECT_ITEM'
const TOGGLE_SELECTION_BAR = 'TOGGLE_SELECTION_BAR'
const SHOW_SELECTION_BAR = 'SHOW_SELECTION_BAR'
const HIDE_SELECTION_BAR = 'HIDE_SELECTION_BAR'

// selectors
export const isSelected = (state, id) =>
  state.selection.selected.find(_id => _id === id) !== undefined
export const getSelectedIds = state => state.selection.selected
export const isSelectionBarVisible = state =>
  state.selection.selected.length !== 0 || state.selection.isSelectionBarOpened
export const getSelectedFiles = state =>
  getSelectedIds(state).map(
    id => getFileById(state, id) || getFileByIdFromCozyClient(state, id)
  )

// actions
export const showSelectionBar = () => ({ type: SHOW_SELECTION_BAR })
export const hideSelectionBar = () => ({ type: HIDE_SELECTION_BAR })
export const toggleSelectionBar = () => ({ type: TOGGLE_SELECTION_BAR })
export const toggleItemSelection = (id, selected) => ({
  type: selected ? UNSELECT_ITEM : SELECT_ITEM,
  id
})

// reducers
const selected = (state = [], action) => {
  if (action.meta && action.meta.cancelSelection) {
    return []
  }
  switch (action.type) {
    case SELECT_ITEM:
      return [...state, action.id]
    case UNSELECT_ITEM: {
      const idx = state.indexOf(action.id)
      return [...state.slice(0, idx), ...state.slice(idx + 1)]
    }
    case HIDE_SELECTION_BAR:
      return []
    default:
      return state
  }
}

const isSelectionBarOpened = (state = false, action) => {
  if (action.meta && action.meta.cancelSelection) {
    return false
  }
  switch (action.type) {
    case TOGGLE_SELECTION_BAR:
      return !state
    case SHOW_SELECTION_BAR:
      return true
    case HIDE_SELECTION_BAR:
      return false
    default:
      return state
  }
}

export default combineReducers({
  selected,
  isSelectionBarOpened
})
