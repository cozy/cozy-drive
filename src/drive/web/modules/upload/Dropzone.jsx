import React, { Component } from 'react'
import { connect } from 'react-redux'
import Dropzone from 'react-dropzone'

import { uploadFiles } from 'drive/web/modules/navigation/duck'

import styles from 'drive/styles/dropzone'
import DropzoneTeaser from 'drive/web/modules/upload/DropzoneTeaser'
class StatefulDropzone extends Component {
  state = {
    dropzoneActive: false
  }

  onDragEnter = evt => {
    if (!canDrop(evt)) return
    this.setState(state => ({ ...state, dropzoneActive: true }))
  }

  onDragLeave = () =>
    this.setState(state => ({ ...state, dropzoneActive: false }))

  onDrop = async (files, _, evt) => {
    const folderId = this.props.displayedFolder.id
    this.setState(state => ({ ...state, dropzoneActive: false }))
    if (!canDrop(evt)) return
    const filesToUpload = canHandleFolders(evt) ? evt.dataTransfer.items : files
    this.props.uploadFiles(filesToUpload, folderId)
  }

  render() {
    const { dropzoneActive } = this.state
    const { displayedFolder, children, ...rest } = this.props
    return (
      <Dropzone
        {...rest}
        className={dropzoneActive ? styles['fil-dropzone-active'] : ''}
        disableClick
        style={{}}
        onDrop={this.onDrop}
        onDragEnter={this.onDragEnter}
        onDragLeave={this.onDragLeave}
      >
        {dropzoneActive && <DropzoneTeaser currentFolder={displayedFolder} />}
        {children}
      </Dropzone>
    )
  }
}

// DnD helpers for folder upload
const canHandleFolders = evt => {
  if (!evt.dataTransfer) return false
  const dt = evt.dataTransfer
  return dt.items && dt.items.length && dt.items[0].webkitGetAsEntry != null
}

const canDrop = evt => {
  const items = evt.dataTransfer.items
  for (let i = 0; i < items.length; i += 1) {
    if (items[i].kind !== 'file') return false
  }
  return true
}

const mapDispatchToProps = dispatch => ({
  uploadFiles: (files, folderId) => {
    dispatch(uploadFiles(files, folderId))
  }
})

export default connect(
  null,
  mapDispatchToProps
)(StatefulDropzone)
