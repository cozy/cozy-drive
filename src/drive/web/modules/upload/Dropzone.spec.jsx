import React from 'react'
import { mount } from 'enzyme'
import { setupFolderContent, mockCozyClientRequestQuery } from 'test/setup'
import Dropzone, { Dropzone as DumbDropzone } from './Dropzone'
import AppLike from 'test/components/AppLike'
import { uploadFiles } from 'drive/web/modules/navigation/duck'

jest.mock('drive/web/modules/navigation/duck', () => ({
  uploadFiles: jest.fn().mockReturnValue({
    type: 'FAKE_UPLOAD_FILES'
  })
}))

mockCozyClientRequestQuery()

describe('Dropzone', () => {
  it('should dispatch the uploadFiles action', async () => {
    const displayedFolder = {
      id: 'directory-foobar0'
    }
    const { store, client } = await setupFolderContent({
      folderId: 'directory-foobar0'
    })
    const root = mount(
      <AppLike client={client} store={store}>
        <Dropzone displayedFolder={displayedFolder} />
      </AppLike>
    )

    const dropzone = root.find(DumbDropzone)
    const files = []
    dropzone.props().uploadFiles(files)
    expect(uploadFiles).toHaveBeenCalledWith(
      files,
      'directory-foobar0',
      expect.objectContaining({
        refresh: expect.any(Function)
      })
    )
  })
})
