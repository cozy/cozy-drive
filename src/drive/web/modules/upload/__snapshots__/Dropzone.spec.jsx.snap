// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropzone should match snapshot 1`] = `
<AppLike
  client={
    CozyClient {
      "uri": undefined,
    }
  }
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "dispatch": [MockFunction],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <CozyProvider
      client={
        CozyClient {
          "uri": undefined,
        }
      }
    >
      <TestI18n>
        <I18n
          defaultLang="en"
          dictRequire={[Function]}
          lang="en"
        >
          <AcceptingSharingProvider>
            <HashRouter>
              <Router
                location={
                  Object {
                    "hash": "",
                    "key": "default",
                    "pathname": "/",
                    "search": "",
                    "state": null,
                  }
                }
                navigationType="POP"
                navigator={
                  Object {
                    "action": "POP",
                    "createHref": [Function],
                    "encodeLocation": [Function],
                    "go": [Function],
                    "listen": [Function],
                    "location": Object {
                      "hash": "",
                      "key": "default",
                      "pathname": "/",
                      "search": "",
                      "state": null,
                    },
                    "push": [Function],
                    "replace": [Function],
                  }
                }
              >
                <ThumbnailSizeContextProvider>
                  <BreakpointsProvider>
                    <PushBannerProvider>
                      <FabProvider>
                        <WithSharingState(withClient(withVaultClient(Connect(Dropzone))))
                          displayedFolder={
                            Object {
                              "id": "directory-foobar0",
                            }
                          }
                        >
                          <withClient(withVaultClient(Connect(Dropzone)))
                            displayedFolder={
                              Object {
                                "id": "directory-foobar0",
                              }
                            }
                            sharingState={
                              Object {
                                "getRecipients": [MockFunction],
                                "getSharingLink": [MockFunction],
                                "hasWriteAccess": [MockFunction],
                                "refresh": [MockFunction],
                              }
                            }
                          >
                            <withVaultClient(Connect(Dropzone))
                              client={
                                CozyClient {
                                  "uri": undefined,
                                }
                              }
                              displayedFolder={
                                Object {
                                  "id": "directory-foobar0",
                                }
                              }
                              sharingState={
                                Object {
                                  "getRecipients": [MockFunction],
                                  "getSharingLink": [MockFunction],
                                  "hasWriteAccess": [MockFunction],
                                  "refresh": [MockFunction],
                                }
                              }
                            />
                          </withClient(withVaultClient(Connect(Dropzone)))>
                        </WithSharingState(withClient(withVaultClient(Connect(Dropzone))))>
                      </FabProvider>
                    </PushBannerProvider>
                  </BreakpointsProvider>
                </ThumbnailSizeContextProvider>
              </Router>
            </HashRouter>
          </AcceptingSharingProvider>
        </I18n>
      </TestI18n>
    </CozyProvider>
  </Provider>
</AppLike>
`;
